File: mongodb-ruby/lib/cerbos/mongodb/guards.rb
# frozen_string_literal: true

require_relative "aggregation"
require_relative "errors"
require_relative "operands"

module Cerbos
  module MongoDB
    # The conjuncts that keep out documents on which CEL raises. CEL denies on an error under
    # BOTH polarities, while a MongoDB filter has no UNKNOWN: a negation over a predicate that
    # could not be evaluated would match. So every guard is ANDed OUTSIDE the predicate — and
    # outside any +$nor+ a negation wraps it in.
    module Guards
      extend Operands

      # Operators whose second operand is a lambda that binds an iteration variable.
      LAMBDA_BINDING = %w[exists exists_one all filter map except].freeze

      module_function

      # ["a", "b"], v → {"a" => {"b" => v}}; an empty path is v itself.
      def field_filter(path, value)
        path.reverse.reduce(value) { |acc, key| {key => acc} }
      end

      # A single-field comparison, with the field required to exist (+require_exists+) and/or to
      # be non-null (+nullable+: a stored null is a missing Cerbos attribute) ANDed before it.
      def guarded_field_filter(path, value, nullable, require_exists = false)
        filter = field_filter(path, value)
        return filter unless nullable || require_exists

        guards = []
        guards << field_filter(path, {"$exists" => true}) if require_exists
        guards << field_filter(path, {"$ne" => nil}) if nullable
        {"$and" => guards + [filter]}
      end

      # ANDs {field: {$ne: null}} before +filter+ for every nullable field the operands read.
      def with_nullable(filter, operands, mapper)
        names = operands.flat_map { |operand| variable_names(operand) }.uniq.select { |name| mapper.nullable?(name) }
        return filter if names.empty?

        {"$and" => names.map { |name| field_filter(mapper.resolve_field(name).path, {"$ne" => nil}) } + [filter]}
      end

      # "Every to-one parent this operand dots through is present", or nil when it dots through
      # none.
      #
      # CEL cannot dot through a list, so every intermediate segment of `a.b.c` is a to-ONE
      # parent: absent, the application sends no attribute and CEL raises a missing-path error,
      # which denies. The flattened path a MongoDB filter matches cannot see that — an absent
      # parent and a childless parent both fail the match — so a +$nor+ over it is TRUE for a
      # parentless document. Requiring the parent OUTSIDE the +$nor+ is faithful under both
      # polarities (cerbos/query-plan-adapters#315, #316).
      #
      # `<parent>.0` exists exactly when the parent array is non-empty, which is how the document
      # model spells "the to-one parent was serialised" for a +requires_parent+ declaration. A
      # +type: :one+ relation needs no declaration: a to-one hop is by definition a level that
      # can be absent, and {path: {$ne: null}} excludes both the absent and the stored-null case
      # (cerbos/query-plan-adapters#375).
      def required_parents(operand, mapper)
        array_parents = []
        to_one_paths = []
        variable_names(operand).each do |name|
          relation = mapper.relation_of(name)
          next if relation.nil?

          array_parents << relation.requires_parent if relation.requires_parent
          to_one_paths << relation.name if relation.type == :one
        end
        clauses = array_parents.uniq.map { |parent| {"#{parent}.0" => {"$exists" => true}} } +
          to_one_paths.uniq.map { |path| {path => {"$ne" => nil}} }
        return nil if clauses.empty?

        (clauses.length == 1) ? clauses.first : {"$and" => clauses}
      end

      # The collection variables +operand+ reads as a list: the collection of a `needle in list`,
      # a hasIntersection operand, and the collection a map or other macro ranges over. Stops at
      # a lambda body, whose references are element-scoped. With +strict_only+, also stops at
      # `&&`, `||` and a ternary, which CEL may evaluate without reading every operand.
      def list_reads(operand, strict_only)
        return [] unless expression?(operand)

        first, second = operand.operands
        case operand.operator
        when "in"
          return (second && variable?(second)) ? [second.name] : []
        when "hasIntersection"
          return operand.operands.flat_map { |child| variable?(child) ? [child.name] : list_reads(child, strict_only) }
        when "and", "or", "if"
          return [] if strict_only
        else
          if LAMBDA_BINDING.include?(operand.operator)
            return (first && variable?(first)) ? [first.name] : []
          end
        end
        operand.operands.flat_map { |child| list_reads(child, strict_only) }
      end

      # "Every list the negated +operand+ reads is stored as an array", or nil when it reads none.
      #
      # A list that is null or absent is a CEL error — `2 in null` has no overload, and an absent
      # attribute is a missing-path error — so check() denies the document under BOTH polarities.
      # MongoDB answers the membership false there instead, which the $nor a negation wraps it in
      # turns into a match. {list: {$type: "array"}} holds for every stored array, the empty one
      # included, and for nothing else, so requiring it OUTSIDE the $nor denies exactly the
      # documents CEL cannot evaluate.
      #
      # Only a membership CEL is certain to evaluate can be guarded: a list read under `&&`, `||`
      # or a ternary may be skipped by CEL, where the guard would deny a document the PDP allows.
      # Those, and a list read inside a collection predicate (whose path is element-relative), are
      # refused.
      def list_shape_guard(operand, mapper, root_scope)
        names = list_reads(operand, false).uniq
        return nil if names.empty?

        always_read = list_reads(operand, true)
        if !root_scope || names.any? { |name| !always_read.include?(name) }
          raise UnsupportedError,
            "a negated membership whose list may go unevaluated, or is element-scoped, cannot be " \
            "guarded against a null or absent list, which CEL denies and a $nor would match"
        end

        paths = names.filter_map { |name|
          resolved = mapper.resolve_field(name)
          relation = resolved.relation
          if relation.nil?
            resolved.path.join(".")
          elsif relation.type == :many && relation.requires_parent.nil?
            # A to-many relation's elements live in one array on the document; a relation that
            # declares its parent is already guarded by required_parents.
            relation.name
          end
        }.uniq
        clauses = paths.map { |path| {path => {"$type" => "array"}} }
        return nil if clauses.empty?

        (clauses.length == 1) ? clauses.first : {"$and" => clauses}
      end

      # Keeps out every document on which an operand of +filter+ could not be evaluated: a
      # nullable field that is null, a guarded aggregation expression, and an absent to-one
      # parent. Each is ANDed OUTSIDE +filter+.
      def with_evaluation(filter, operands, mapper)
        guarded = with_nullable(filter, operands, mapper)
        conjuncts = operands.flat_map { |operand| Aggregation.evaluation_guards(operand, mapper) } +
          operands.filter_map { |operand| required_parents(operand, mapper) }
        conjuncts.empty? ? guarded : {"$and" => conjuncts + [guarded]}
      end
    end
  end
end

File: mongoose/src/guards.ts
import type { PlanExpressionOperand } from "@cerbos/core";

import { buildEvaluationGuards } from "./aggregation";
import type { Mapper, MongooseFilter } from "./index";
import {
  isNullableReference,
  relationOfReference,
  resolveFieldReference,
} from "./mapper";
import { collectVariableNames } from "./operands";

/** `["a", "b"]`, `v` → `{ a: { b: v } }`; an empty path is `v` itself. */
export const buildFieldFilter = (path: string[], value: unknown): any =>
  path.reduceRight((acc: unknown, key) => ({ [key]: acc }), value);

/**
 * A single-field comparison, with the field required to exist (`requireExists`) and/or to be
 * non-null (`nullable`: a stored null is a missing Cerbos attribute) ANDed in front of it.
 */
export const buildGuardedFieldFilter = (
  path: string[],
  value: unknown,
  nullable: boolean,
  requireExists = false,
): MongooseFilter => {
  const filter = buildFieldFilter(path, value);
  if (!nullable && !requireExists) {
    return filter;
  }
  const guards: MongooseFilter[] = [];
  if (requireExists) {
    guards.push(buildFieldFilter(path, { $exists: true }));
  }
  if (nullable) {
    guards.push(buildFieldFilter(path, { $ne: null }));
  }
  return {
    $and: [...guards, filter],
  };
};

/** ANDs `{ field: { $ne: null } }` in front of `filter` for every nullable field the operands read. */
export const withNullableGuards = (
  filter: MongooseFilter,
  operands: PlanExpressionOperand[],
  mapper: Mapper,
): MongooseFilter => {
  const guardedNames = [
    ...new Set(
      operands
        .flatMap(collectVariableNames)
        .filter((name) => isNullableReference(name, mapper)),
    ),
  ];
  if (guardedNames.length === 0) {
    return filter;
  }

  const guards = guardedNames.map((name) => {
    const { path } = resolveFieldReference(name, mapper);
    return buildFieldFilter(path, { $ne: null });
  });
  return { $and: [...guards, filter] };
};

/**
 * "Every to-one parent this expression dots through is present", as a Mongoose filter, or
 * undefined when it dots through none.
 *
 * CEL cannot dot through a list, so each intermediate segment of `a.b.c` is a to-ONE parent:
 * absent, the application sends no attribute at all and CEL raises a missing-path error,
 * which denies. The flattened `a.b` path a Mongo filter matches against cannot see that — an
 * absent parent and a childless parent both fail the match — so any `$nor` over it is TRUE
 * for a parentless document and returns rows the PDP denies.
 *
 * `requiresParent` already carried this for the `$size` aggregation (#309); membership,
 * `hasIntersection` and the negated count spelling reach the same chain without ever passing
 * through it (cerbos/query-plan-adapters#315, #316). Requiring the parent OUTSIDE the `$nor`
 * fixes all of them at once, and is unconditionally faithful to CEL rather than a
 * per-operator patch: a missing parent denies under BOTH polarities regardless of which
 * operator sits above the chain.
 *
 * `<parent>.0` exists exactly when the parent array is non-empty, which is how the document
 * model spells "the to-one parent was serialised" — the same test the `$size` guard makes
 * with `$ifNull`. That `$size` guard alone is not enough: it makes the count `null`, and BSON
 * orders `null` BELOW every number, so `$eq`/`$gt`/`$gte` against it are false but `$lte`/`$lt`
 * are TRUE — `size(chain) <= 0` admitted every parentless document. A filter-level conjunct
 * has no such ordering to get wrong.
 *
 * `requiresParent` only ever appears on a top-level mapping, so the paths produced here are
 * always rooted at the document. A nested `fields` entry that declared one would need its
 * path rebased onto the enclosing `$elemMatch` scope instead.
 *
 * A `type: "one"` relation needs no `requiresParent` to opt in: a to-one hop is BY DEFINITION
 * a level that can be absent, and reaching a scalar through one is a missing-path error when it
 * is. `{ path: { $ne: null } }` is the subdocument spelling of the same requirement — a missing
 * path compares equal to null in MongoDB, so it excludes both the absent and the stored-null
 * case, and on a two-level path it subsumes the level above (cerbos/query-plan-adapters#375).
 */
const buildRequiredParentsFilter = (
  operand: PlanExpressionOperand,
  mapper: Mapper,
): MongooseFilter | undefined => {
  const arrayParents = new Set<string>();
  const toOnePaths = new Set<string>();
  for (const name of collectVariableNames(operand)) {
    const relation = relationOfReference(name, mapper);
    if (relation?.requiresParent !== undefined) {
      arrayParents.add(relation.requiresParent);
    }
    if (relation?.type === "one") {
      toOnePaths.add(relation.name);
    }
  }
  const clauses: MongooseFilter[] = [
    ...[...arrayParents].map((parentPath) => ({
      [`${parentPath}.0`]: { $exists: true },
    })),
    ...[...toOnePaths].map((path) => ({ [path]: { $ne: null } })),
  ];
  if (clauses.length === 0) {
    return undefined;
  }
  return clauses.length === 1 ? clauses[0]! : { $and: clauses };
};

/**
 * Keeps out every document on which an operand of `filter` could not be evaluated: a nullable
 * field that is null, a guarded expression (see `AggregationOperator.guard`) that cannot be
 * evaluated, and an absent to-one parent. Each is ANDed OUTSIDE `filter`.
 */
export const withEvaluationGuards = (
  filter: MongooseFilter,
  operands: PlanExpressionOperand[],
  mapper: Mapper,
): MongooseFilter => {
  const guardedFilter = withNullableGuards(filter, operands, mapper);
  const guards = operands.flatMap((operand) =>
    buildEvaluationGuards(operand, mapper),
  );
  // The absent to-one parent is the other way an operand can be undefined, and it belongs
  // here with the rest: a filter-level conjunct applies to whatever `filter` already is, so
  // both the plain comparison and the `$nor` a negation wraps it in inherit it (#315, #316).
  const requiredParents = operands
    .map((operand) => buildRequiredParentsFilter(operand, mapper))
    .filter((parents): parents is MongooseFilter => parents !== undefined);
  const conjuncts = [...guards, ...requiredParents];
  return conjuncts.length === 0
    ? guardedFilter
    : { $and: [...conjuncts, guardedFilter] };
};
